### ISSUE TITLE
Upgrade NPC Farmers from Roombas to a better version 

### ISSUE THREAD

**Issue by Klaranth** [pandorabox-io/in-game#299]
AerfudaXR 
Can it be made so npc can take any crops they walk instantly, because i wait long and when they stand or cross the crops, they sometimes wont pick and it becomes hard
OgelGames 
Farming is a slow process, npcs can't make crops grow faster, and they won't pick unripe crops

AerfudaXAlt 
I mean on ripen crops
So i put npc into a field, it ran through ripen crops, but sometimes it just pass and the other time it picked, why not make so it can harvest whenever stumble upon a ripen crops immediately instead of only passing by

1hit
npcs are like really old roombas; random walks and pick up what they touch?

AerfudaXAlt 
Yes, what the hitbox touches
It is randomized, but if you know the function of farmers
Well, they usually place on small fields, but what do you think the use of a farmer if they are only used in small fields?

**Comment by SwissalpS** [comment]
I have no trouble with them, I tell them to stand still and they take care of the 3x3 square around them neatly and eficiently :)

**Comment by S-S-X** [comment]
NPCs also cannot really find stuff, reason why most mobs feel so stupid is that path finder is disabled here https://github.com/pandorabox-io/pandorabox_custom/blob/c37da38ab8bb53ad18e2cd80f6ec8aca32142214/hacks.lua#L3-L8

Farmers could use custom simplified method for areas without too many obstacles but I'd recommend against reinventing wheel: only very simple method that will just stop if real path finder would be needed.
Basically that is what mobs already do without path finder: just attempt to walk directly towards target and give up if it did not work.

Some simple intelligence for farmers would be useful to predict when and where there's something to do without checking large area, some primitive "vision" could be simulated without actually checking state of world very often.

**Comment by OgelGames** [comment]
I added some code to the farmers to check the node they are standing in, and also widened their search area from 3x3 to 5x5. That should fix the "walking over ripe crops and doing nothing".

> NPCs also cannot really find stuff, reason why most mobs feel so stupid is that path finder is disabled here https://github.com/pandorabox-io/pandorabox_custom/blob/c37da38ab8bb53ad18e2cd80f6ec8aca32142214/hacks.lua#L3-L8

Actually the pathfinder is only used when attacking something, the "stay near" code just searches for nodes and points them in the direction:
https://github.com/minetest-mirrors/mobs_redo/blob/5c9e057ad1a3b0237c39c33303a728101daf1148/api.lua#L627-L659
